orttraining/orttraining/python/training/pytorch_trainer_options.py

```diff
@@ -0,0 +1,376 @@
+            specifies learning rate scheduler
+        mixed_precision (dict):
+            mixed precision training options
+        mixed_precision.enabled (bool, default is False):
+            enable mixed precision (fp16)
+        mixed_precision.loss_scaler (amp.LossScaler, default is None):
+            specifies a loss scaler to be used for fp16. If not specified,
+            :py:class:`.DynamicLossScaler` is used with default values.
+            Users can also instantiate :py:class:`.DynamicLossScaler` and
+            override its parameters. Lastly, a completely new implementation
+            can be specified by extending :py:class:`.LossScaler` class from scratch
+        utils (dict):
+            miscellaneous options
+        utils.grad_norm_clip (bool, default is False):
+            enables gradient norm clipping for 'AdamOptimizer' and 'LambOptimizer'
+        _internal_use (dict):
+            internal, possibly undocumented, options that might be removed in the next release
+        _internal_use.frozen_weights (list, []):
+            list of model parameters to freeze (stop training)
+        _internal_use.enable_internal_postprocess (bool, default is True):
+            enable internal internal post processing of the ONNX model
+        _internal_use.extra_postprocess (callable, default is None)
+            a functor to postprocess the ONNX model.
+            It does not override :py:attr:`._internal_use.enable_internal_postprocess`, but complement it
+
+    Example:
+        .. code-block:: python
+
+            opts = PytorchTrainerOptions({
+                               'batch' : {
+                                   'gradient_accumulation_steps' : 128
+                               },
+                               'device' : {
+                                   'id' : 'cuda:0',
+                                   'mem_limit' : 2*1024*1024*1024,
+                               },
+                               'lr_scheduler' : optim.lr_scheduler.LinearWarmupLRScheduler(),
+                               'mixed_precision' : {
+                                   'enabled': True,
+                                   'loss_scaler': amp.LossScaler(loss_scale=float(1 << 16))
+                               }
+            })
+            fp16_enabled = opts.mixed_precision.enabled
+     """
+
+    def __init__(self, options, _validate=True):
+        # Keep a copy of original input for debug
+        self._original_opts = dict(options)
+
+        # Add an empty dictionary for non specified nested dicts
+        subgroups = [k for k, v in _PYTORCH_TRAINER_OPTIONS_SCHEMA.items()
+                     if isinstance(v, dict) and 'type' in v and v['type'] == 'dict']
+        self._validated_opts = dict(self._original_opts)
+        if _validate:
+            for name in subgroups:
+                if name not in self._validated_opts:
+                    self._validated_opts[name] = {}
```

Comment:
why do we need to manually set empty for `dict` type options not provided, but not other type of options?

Reply:
That is a limitation on cerberus lib. `dict` doesn't have a default setter to `{}`. Without an empty `{}`, all the internal default values are not populated. I may have found an official solution for that, which I will test today

Reply:
I have fixed this!